tokio/src/io/async_fd.rs

```diff
@@ -0,0 +1,269 @@
+///        let mut result = Poll::Pending;
+///        while result.is_pending() {
+///            // Poll the saved future; if it's not ready, our context waker will be saved in the
+///            // future and we can return.
+///            ready!(self.poller.as_mut().poll(cx));
+///
+///            // Reset the poller future, since we consumed it.
+///            let arc = self.async_fd.clone();
+///            self.poller = Box::pin(async move {
+///                let _ = arc.readable().await.map(|mut guard| guard.retain_ready());
+///            });
+///
+///            // Because we need to bind the ReadyGuard to the lifetime of self, we have to re-poll here.
+///            // It's possible that we might race with another thread clearing the ready state, so deal
+///            // with that as well.
+///            let fut = self.async_fd.readable();
+///            tokio::pin!(fut);
+///            result = fut.as_mut().poll(cx);
+///        }
+///
+///        result
+///    }
+/// }
+/// ```
+///
+/// [`readable`]: method@Self::readable
+/// [`writable`]: method@Self::writable
+/// [`ReadyGuard`]: struct@self::ReadyGuard
+/// [`TcpStream::poll_read_ready`]: struct@crate::net::TcpStream
+pub struct AsyncFd {
+    handle: Handle,
+    fd: RawFd,
+    shared: slab::Ref<ScheduledIo>,
+}
+
+impl std::fmt::Debug for AsyncFd {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AsyncFd").field("fd", &self.fd).finish()
+    }
+}
+
+unsafe impl Send for AsyncFd {}
+unsafe impl Sync for AsyncFd {}
+
+const fn all_interest() -> mio::Interest {
+    mio::Interest::READABLE.add(mio::Interest::WRITABLE)
+}
+
+/// Represents an IO-ready event detected on a particular file descriptor, which
+/// has not yet been acknowledged. This is a `must_use` structure to help ensure
+/// that you do not forget to explicitly clear (or not clear) the event.
+#[must_use = "You must explicitly choose whether to clear the readiness state by calling a method on ReadyGuard"]
+pub struct ReadyGuard<'a> {
+    async_fd: &'a AsyncFd,
+    event: Option<ReadyEvent>,
+}
+
+impl<'a> std::fmt::Debug for ReadyGuard<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ClearReady")
+            .field("async_fd", self.async_fd)
+            .finish()
+    }
+}
+
+impl<'a> ReadyGuard<'a> {
+    /// Indicates to tokio that the file descriptor is no longer ready. The
+    /// internal readiness flag will be cleared, and tokio will wait for the
+    /// next edge-triggered readiness notification from the OS.
+    ///
+    /// It is critical that this function not be called unless your code
+    /// _actually observes_ that the file descriptor is _not_ ready. Do not call
+    /// it simply because, for example, a read succeeded; it should be called
+    /// when a read is observed to block.
+    ///
+    /// [`drop`]: method@std::mem::drop
+    pub fn clear_ready(&mut self) {
```

Comment:
Since both `with_io` and `with_poll` will clear readiness automatically, what use case do you imagine these methods fulfilling? Grabbing a `ReadyGuard`, and then using the FD separately?

Reply:
Potentially using the fd with some external library that isn't adapted to returning `io::Result`s, for example.